learn/develop/sparse-data/index.qmd

```diff
@@ -0,0 +1,92 @@
+```
+The sparse representation of this vector only requires 5 values. 1 value for the length (25), 2 values for the locations of the non-zero values (1, 22), and 2 values for the non-zero values (100, 1). This idea can also be extended to matrices as is done in the Matrix package.
+
+The Matrix package implements sparse vectors and sparse matrices as S4 objects. These objects are the go-to objects in the R ecosystem for dealing with sparse data.
+
+The [tibble](https://tibble.tidyverse.org/) is used as the main carrier of data inside tidymodels. While matrices or data frames are accepted as input in different places, they are converted to tibbles early to take advantage of the benefits and checks they provide. This is where the [sparsevctrs](https://github.com/r-lib/sparsevctrs) package comes in. The sparse vectors and matrices from the Matrix package don't work with data frames or tibbles and are thus not able to be used well within tidymodels. The sparsevctrs package allows for the creation of [ALTREP](https://svn.r-project.org/R/branches/ALTREP/ALTREP.html) vectors that act like normal numeric vectors but are encoded sparsely. If an operation on the vector can't be done sparsely, it will fall back and **materialize**. Materialization means that it will generate and cache a dense version of the vector that is then used. 
+
+These sparse vectors can be put into tibbles and are what allow tidymodels to handle sparse data. We will henceforth refer to a tibble that contains sparse columns as a **sparse tibble**. Sparse matrices require all the elements to be of the same type. All logical, all integer, or all doubles. This limitation does not apply to sparse tibbles as you can store both sparse vectors and dense vectors. But also mix and match numeric, factors, datetimes, and more.
+
+The sparsity mostly matters for the predictors. The outcomes, case weights, and predictions are done densely.
+
+Below is outlined how sparse data (matrix & tibble) works with the various tidymodels packages.
+
+## rsample
+
+The resampling functions from the rsample package work with sparse tibbles out of the box. However, they won't work with sparse matrices. Instead, use the `coerce_to_sparse_tibble()` function from the sparsevctrs package to turn the sparse matrix into a sparse tibble and proceed with that object.
+
+```r
+library(sparsevctrs)
+data_tbl <- coerce_to_sparse_tibble(data_mat)
+```
+
+## recipes
+
+The recipes package receives data in `recipe()`, `prep()`, and `bake()`. These functions all handle sparse data in the same manner.
+
+Sparse tibbles should work as normal. Sparse matrices are accepted and then turned into sparse tibbles right away, and flow through the internals as sparse tibbles.
+
+The `composition` argument of `bake()` understands sparse tibbles. When `composition = "dgCMatrix"` the resulting sparse matrix will be created from the sparse tibble with minimal overhead.
```

Comment:
I assume some sort of `model.matrix()`-ey thing needs to happen for the sparse tibble -> `dgCMatrix` conversion? What happens to e.g. factors? I guess I'm surprised to hear that this conversion is easy given that the dense analogous (tibble to matrix) is so painful.

Reply:
the sparse tibble to dgCMatrix conversion requires all columns are numeric, otherwise it will error. This was also the past behavior.

Reply:
I see!